subsys/net/ip/tcp.c

```diff
@@ -3252,72 +3294,65 @@ int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 		goto out;
 	}
 
-	len = net_pkt_get_len(pkt);
+	if (msg) {
+		len = 0;
 
-	if (conn->send_data->buffer) {
-		orig_buf = net_buf_frag_last(conn->send_data->buffer);
+		for (int i = 0; i < msg->msg_iovlen; i++) {
+			len += msg->msg_iov[i].iov_len;
+		}
 	}
 
-	net_pkt_append_buffer(conn->send_data, pkt->buffer);
-	conn->send_data_total += len;
-	NET_DBG("conn: %p Queued %zu bytes (total %zu)", conn, len,
-		conn->send_data_total);
-	pkt->buffer = NULL;
+	/* Queue no more than TX window permits. */
+	len = MIN(conn->send_win - conn->send_data_total, len);
```

Comment:
What happens of send_win < send_data_total. I think here is a potential for an unintended wrap when the send_win decreases during runtime.

Reply:
The diff is poorly formatted so it's not clearly visible - the `tcp_window_full()` check just a few lines above verifies that `send_data_total < send_win` (otherwise we have a full window case and quit), and as we hold the connection mutex, neither of the two variables should change in between.

Reply:
I see it could not go wrong for that reason,, I think it would be better to add a comment that this edge case is already covered. Somenoe modifying this code later on might not see this relation.

Reply:
Sure - I've elaborated in the comment why it's safe to do the subtraction.